File: participant.py
import grpc
import time
import random
import sys
import os
from concurrent import futures

import twopc_pb2
import twopc_pb2_grpc

SEPARATOR = "-" * 60

import builtins
print = lambda *args, **kwargs: builtins.print(*args, flush=True, **kwargs)

class VotingParticipant(twopc_pb2_grpc.VotingServiceServicer):
    def Vote(self, request, context):
        node_id = os.getenv("NODE_ID", "unknown")
        print(f"Phase Voting of Node {node_id} receives RPC Vote from Phase Voting of Node coordinator")

        commit_decision = random.choice([True, False])
        decision_str = "Commit" if commit_decision else "Abort"

        print(f"[{node_id}] Voting {decision_str} for Transaction {request.transaction_id}")
        print(SEPARATOR)

        return twopc_pb2.VoteResponse(
            transaction_id=request.transaction_id,
            node_id=node_id,
            commit=commit_decision
        )
def serve():
    port = os.getenv("PORT", "50051")
    node_id = os.getenv("NODE_ID", "unknown")

    print(f"[{node_id}] Phase Voting of Node {node_id} listening on port {port}...")

    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    twopc_pb2_grpc.add_VotingServiceServicer_to_server(VotingParticipant(), server)
    server.add_insecure_port(f"[::]:{port}")
    server.start()
    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        print(f"[{node_id}] Shutting down service...")
        server.stop(0)
    except Exception as e:
        print(f"[{node_id} ERROR] {e}", file=sys.stderr)
if __name__ == "__main__":
    serve()

File: participant.go
package main

import (
	"context"
	"log"
	"net"
	"os"
	"time"

	pb "twopcapp/proto"
	"google.golang.org/grpc"
)

const Divider = "------------------------------------------------"

type ParticipantServer struct {
	pb.UnimplementedDecisionServiceServer
	pb.UnimplementedGlobalDecisionServiceServer
	nodeID   string
	lastVote bool
}

func (s *ParticipantServer) ReportVote(ctx context.Context, req *pb.VoteReport) (*pb.Ack, error) {
	log.Printf("[%s] Phase Decision of Node %s received ReportVote from Node %s: Commit=%v",
		time.Now().Format(time.RFC3339Nano), s.nodeID, req.NodeId, req.Commit)
	s.lastVote = req.Commit
	return &pb.Ack{Message: "Vote recorded"}, nil
}

func (s *ParticipantServer) GlobalDecision(ctx context.Context, req *pb.DecisionMessage) (*pb.Ack, error) {
	outcome := "ABORTED"
	if req.GlobalCommit {
		outcome = "COMMITTED"
	}

	log.Printf("\n%s\n[%s] Phase Decision of Node %s received GlobalDecision for Tx %s => %s\n%s",
		Divider,
		time.Now().Format(time.RFC3339Nano), s.nodeID, req.TransactionId, outcome,
		Divider,
	)

	// Log the local action taken based on the global decision
	if req.GlobalCommit {
		log.Printf("[Participant %s] Locally COMMIT the transaction %s", s.nodeID, req.TransactionId)
	} else {
		log.Printf("[Participant %s] Locally ABORT the transaction %s", s.nodeID, req.TransactionId)
	}

	return &pb.Ack{Message: "Decision applied"}, nil
}

func main() {
	log.SetFlags(0)
	log.SetOutput(os.Stdout)

	nodeID := os.Getenv("NODE_ID")
	if nodeID == "" {
		log.Fatal("NODE_ID environment variable not set")
	}

	port := ":60051"
	lis, err := net.Listen("tcp", port)
	if err != nil {
		log.Fatalf("Failed to listen: %v", err)
	}

	grpcServer := grpc.NewServer()
	server := &ParticipantServer{nodeID: nodeID}
	pb.RegisterDecisionServiceServer(grpcServer, server)
	pb.RegisterGlobalDecisionServiceServer(grpcServer, server)

	log.Printf("[%s] Phase Decision of Node %s listening on %s...",
		time.Now().Format(time.RFC3339Nano), nodeID, port)

	if err := grpcServer.Serve(lis); err != nil {
		log.Fatalf("Failed to serve: %v", err)
	}
}
